Extract the Unix socket command system from the application

nemanja-boric-sociomantic opened the pull request:
The only convenient way to use the Unix domain socket as a control point
was through the `UnixSocketExt`, which was limiting this to a single
socket per application. Now the `UnixSocketExt` is only instantiating and
configuring the instance of `UnixSocketControlInterface`, which can
also be instantiated and configured manually as many times as needed.

nemanja-boric-sociomantic committed:
Extract the Unix socket command system from the application

The only convenient way to use the Unix domain socket as a control point
was through the `UnixSocketExt`, which was limiting this to a single
socket per application. Now the `UnixSocketExt` is only instantiating
and configuring the instance of `UnixListener` and `CommandsRegistry`,
which can also be instantiated and configured manually as many times as
needed.

Fixes #523

nemanja-boric-sociomantic committed:
Accept UnixSocketListener's path as a cstring

gavin-norman-sociomantic commented on src/ocean/util/UnixSocketCommandInterface.d:
```diff
@@ -0,0 +1,161 @@
+/*******************************************************************************
```
I question the choice of `util` for this module. As it's inextricably connected with `ocean.net.server.unix.UnixListener`, wouldn't it make more sense to put it in the same package? (I know it's not actually related to network I/O, but the same goes for loads of the stuff in `ocean.net.http`, for example.)

nemanja-boric-sociomantic replied:
It didn't feel all right for me to be in `ocean.net.*`, since it does not have much to do with the network itself - this just presents itself as way to add and remove the handlers and run them. And it's really the utility module IMHO. I somehow feel the difference between `http` and this - while the former is full blown http server, this is just a small utility.

I'm open for any other proposals, though, as I don't feel strongly about this.

gavin-norman-sociomantic commented on src/ocean/util/UnixSocketCommandInterface.d:
```diff
@@ -0,0 +1,161 @@
+import ocean.io.select.EpollSelectDispatcher;
+
+/// ditto
+public class UnixSocketCommandInterface
```
Any reason for this to be a class?

nemanja-boric-sociomantic replied:
No, good catch :+1:

nemanja-boric-sociomantic replied:
Sad, the bit with the connection handlers needs to be a class, since the `UnixListener` takes a reference of the current instance. Making this a struct would copy it and `addHandler`/`removeHandler` would no longer reflect on the actual connection handler.

gavin-norman-sociomantic commented on src/ocean/util/UnixSocketCommandInterface.d:
```diff
@@ -0,0 +1,161 @@
+        Params:
+            epoll = Epoll instance.
+            path = path to bind the socket to
+            mode = mode to apply to the socket
```
It'd be nice to include a link or documentation for which values are possible and what they mean.

nemanja-boric-sociomantic replied:
I'll explain this better - this is just a normal UNIX permission mode (and I default to the reasonable value - `0600`).

gavin-norman-sociomantic commented on src/ocean/util/UnixSocketCommandInterface.d:
```diff
@@ -0,0 +1,161 @@
+
+    /***************************************************************************
+
+        Receives all commands from the socket and splits the command by " ".
```
"all commands" makes it sound like the string can contain multiple commands. The code doesn't look like that, though.

nemanja-boric-sociomantic replied:
Yeah, that's the copy/paste from the old comment.

gavin-norman-sociomantic commented on src/ocean/util/UnixSocketCommandInterface.d:
```diff
@@ -0,0 +1,161 @@
+
+    ***************************************************************************/
+
+    public void listen ( EpollSelectDispatcher epoll,
```
I feel slightly uncomfortable about this class being basically about string parsing and handler lookup, but then this one method (ok, `shutdown` too) involves networking. Any thoughts?

nemanja-boric-sociomantic replied:
So, the exiting API of the `UnixSocketListener` is that it accepts some object that has a `handle` method and it calls that method to handle the request. So, that bit of the class (along with the `addHandler`, etc) can surely be moved to a separate struct, and from there we can go two ways:

1. Either we still maintain `UnixSocketCommandInterface` as a convenient way to create a socket and handle the commands (we just instantiate), or
2. We just leave the handler bits. However, then user falls into a trap where they need to figure out, honestly, not so trivial `UnixSocketListener`'s API and how to plug these bits together. 

Over these two, I would prefer 1, but I honestly don't see downsides to keep everything in a single struct. However, I was wrong before about the modularization (see the reason for this PR as a reference), so I am keen to change things.

nemanja-boric-sociomantic replied:
I've implemented the approach described in 1 - I've split the handler's registry functionality into a separate class, and created the handy helper to combine the `UnixListener` and `CommandsRegistry` API.

gavin-norman-sociomantic commented on src/ocean/util/UnixSocketCommandInterface.d:
```diff
@@ -0,0 +1,219 @@
+        this.unix_listener =
+            new typeof(this.unix_listener)(this.path, epoll, this.commands_registry);
+
+        if (this.mode >= 0)
```
Could it make any sense to move this mode setting into `UnixSocketListener`? If that were done, I think we could do away with `UnixSocketCommandInterface`. Usage would look like:
```D
auto commands = new CommandsRegistry;
auto unix_listener = new UnixListener(this.path, mode, epoll, commands);
epoll.register(unix_listener);
```

nemanja-boric-sociomantic replied:
Yeah good idea, it is a property of unix socket and it should be there.

nemanja-boric-sociomantic committed:
Accept the mode for the unix socket inside UnixListner

Constructors of `UnixListener` and `UnixSocketListener` now accept
the optional `mode` parameter with the mode to apply on the socket
after creation, if needed. Normally the socket will be created
with 002 umask (so it will default to rw-rw-r-- which should be
enough for everybody), but that can now be overriden.

nemanja-boric-sociomantic commented:
Rebased on top of v4.x.x, changed milestone and addressed comments.

nemanja-boric-sociomantic committed:
Deprecate UnixSocketExt.add/removeInteractiveHandler

There's no need to separate addHandler/removeHandler for
different types of handlers, instead overloads of `addHandler`/
`removeHandler` should be used.

mihails-strasuns-sociomantic commented on src/ocean/net/server/unix/CommandRegistry.d:
```diff
@@ -0,0 +1,134 @@
+        {
+            split(args, " ", this.args_buf);
+            scope predicate = (cstring v) { return v.length; };
+            auto arguments = filter(this.args_buf[], predicate, this.args_buf);
```
This will allocate on each unix socket command, surely we don't want that? https://github.com/sociomantic-tsunami/ocean/blob/v3.x.x/src/ocean/core/array/Transformation.d#L375

Same for filter/split calls few lines above

nemanja-boric-sociomantic replied:
Ah good catch! This is old code so I didn't pay attention.

nemanja-boric-sociomantic replied:
Very nice how this method actually receives a buffer but then ruins it all :-)

mihails-strasuns-sociomantic reviewed (commented):
Very nice! One memory allocation remark though

gavin-norman-sociomantic commented on relnotes/unixcommand.feature.md:
```diff
@@ -0,0 +1,7 @@
+### Extract the Unix socket command system from the application
+
+The only convenient way to use the Unix domain socket as a control point was
```
Commit message needs to be updated. (Mentions `UnixSocketControlInterface`.)
